### DISCUSSION TITLE
布署时,yaml格式错误造成的问题, 以及按键侦测问题与建议

### DISCUSSION THREAD

**Discussion by kevinfis** [rime/home#963] [Q&A]
有几个问题想请教:
1. 建议如果yaml档有错误, 可不可以把错误讯息, 多一点讯息提示?
因为之前我想清除多余方案,
却发现移除 ./build 之后按下部署只能打出英文,
./build删除后没有自动重建立,
或者改每页选项数字后显示结果没变:   "menu/page_size": 5  #例如 5改8, 结果显示结果没变
查了很久一段时间才发现问题出在 标点符号的yaml设定档 有一行没有对齐:
` symbols:`
` # 刚好这一项前面注解很多行`
         '`jh': [ ■ ] ` # 这一行没有对齐好前面, 多了一个空格, 对齐后才解决

布署结果: 没有出现错误仍然可以正常打字,
不过这段期间就是觉得有些设定明明改了却没有生效,
可是因为布署后没错误讯息出现, 不知道如何找原因.
所以才建议多点提示, 如果布署成功但有些功能失效了可以提示一下.

2. 建议功能, 开放使用者设定按下某些组合键后或特定 key symbol时不要动作吗?
因为作笔记方便目前使用emacs,
如果搭配rime输入时, 非常常用的前缀组合键例如 Ctrl-x + b 切换缓冲页 会变成:
 1) 按下C-x: rime 没设定时忽略不动作.  2) 按下b: rime处理器侦测到按键码b, 出现选字选单.
由于按下Ctrl+x Ctrl+s 时rime其实都是忽略不处理的,
实际上应该可以让使用者设定两组按键不动作的吧?
(2组就够用不需要用太多组. 按超过两组按键都太不人性化,也不好记,另外有包解决)

不过如果不能增加这类功能,
我想请教 使用lua达成做法与测试的大概流程,
有可能做到按下某组按键例如 ctrl_L或Alt_R前缀键后, 切换到ascii模式、
之后按下1到2个键后再切换回中文模式吗?

因为还没掌握lua, 只看过arch wiki建立过.xkb设定档用xkbcomp改过按键布局,
目前看rime.lua理解到的代码的是不是大概这样写来侦测按键输入呢?
 ` key:repr() == "Alt_R+x"`
这里使用 keysymdef.h里面的名称格式来判断按键按下再丢给rime进行切换就可以了?

(虽然有emacs-rime不过光是习惯按键用elisp达到的修改难度可能跟改ibus-rime差不多吧?
 目前我rime是设定成单独按下alt_r, 英文键送出:
` editor:    bindings: Control+Return: commit_script_text # 上屏变换后输入`
` key_binder:    - { when: has_menu, accept: "Alt_R", send: "Control+Return" }`
不过使用仍然很不流畅.
加上目前找到的emacs按键设定都是alt组合键的设定, 没有单独设定alt_r触发功能的.
如果有人知道怎么用elisp达到 "Alt_R" 自己决定何时手动切换英文输出, 
作学语言的笔记不用常常输入法切换来切换去.
我会很感谢告知, 那我就用emacs-rime来避免组合键问题就好, 不用折腾lua.)

**Comment by lotem** [comment]
1. 如果YAML語法錯誤，會在RIME的錯誤日誌裏指出錯誤的文件和位置。但如果是合法的YAML文件但定義在了別的地方，目前還沒辦法做更嚴格的檢查。程序只有讀到了錯誤類型或者無法解析的數據才會才會報錯，不會檢查程序不主動讀取的YAML配置是不是寫錯位置了。因爲無法窮盡有效配置的寫法。RIME的配置語法允許用戶自己寫一些YAML數據並在其他位置引用。

**Comment by lotem** [comment]
2. 這個也比較難實現。EMACS的按鍵綁定也是高度可定制的，RIME無法知道哪些按鍵序列在EMACS裏被綁定。要完美解決這個問題，一種方案是根據用戶的EMACS配置在輸入法裏設一遍要迴避哪些按鍵序列，當然還需要在RIME裏開發相應的程序；另一種方案是不使用系統輸入法而是在EMACS裏集成RIME輸入法引擎，如 
DogLooksGood/emacs-rime

**Comment by lotem** [comment]
我本人是手動切換輸入法的。因爲我用的輸入方案也用到Control鍵做輸入碼，更與EMACS全面衝突，只有手動切換輸入模式。

lua 綁定我不太熟悉。可以到該插件的社區諮詢。

**Comment by kevinfis** [comment]
昨天查明我的錯誤之外,
我的第一個問題也有可能不一定算ibus-rime的問題, 
最終試驗的結果,
原本懷疑是方案的yaml 有錯誤不會給出錯誤訊息,
現在更新系統後測試的git版與pacman版發現共同問題狀況是: 
  如果  default.custom.yaml裡面的   schema_list: 裡面的方案名稱不存在,
  佈署後會報錯且系統訊息會提示建立錯誤檔,
  不過/tmp還是看不到 rime.ibus..log.ERROR等檔案,
  只能看到三個連結檔 rime.ibus.ERROR rime.ibus.INFO rime.ibus.WARNING
  但是都打不開.
  而如果方案內容的.schema.yaml裡面有錯, (處理器名稱translator改成 tr這樣的錯誤)
  系統的 notification完全不會出現錯誤, 只會出現 rime is ready.
  
  接著在指令  ibus-daemon -dr 後
  才會出現附有時間的錯誤訊息的三個檔案在 /tmp底下. 
  所以昨天看到的出現錯誤訊息是這樣出現的

因爲這也有可能是ibus或作業系統的問題, 
能夠找到怎麼讓錯誤訊息出現也夠了
我已學到不該佔用版面討論不相關的事,
所以就單純留個訊息在這裡, 
萬一有人用archlinux剛好碰到ibus-rime 1.5.0-1無法產生錯誤訊息時搜尋到作參考,
其他rime額外的功能也會改去lua倉庫那裡提問的.
